Add support for BBCode code style

blurymind opened the pull request:
This makes it possible to do [bb][/bb] code tag style.
To use bbcode style, add this line to the style:
```
	"default": {
		tagStyle: "bbcode"
	},
```

blurymind commented:
Now it can also do most bbcode tags out of the box when using the `tagStyle: ["[", "]"]` without any regressions to the old functionality. They can also take parameters like `[color=red]red text[/color]`

tleunen commented:
This is very interesting. I'm not against using a different syntax for this.
I wonder if in the future, we should focus on a single syntax instead of supporting more than one.

Any reason why you prefer the bracket syntax in comparison to `<>`?

blurymind commented:
I actually forked this project in order to implement a bbcode parser for
pixijs. There is currently not a single one out there. The reason I needed
it for pixi is gdevelop. I am writing a bbcode object for it l and it uses
pixi for rendering

https://github.com/4ian/GDevelop/pull/1247

Why bbcode? It's more user friendly for non programmers and I am also using
if for yarn, which is now bundled with gdevelop

https://github.com/YarnSpinnerTool/YarnEditor

I am the main developer of yarn now :)

Anyways, I like the library for its flexibility. When forking it, I did
this in the spirit of retaining that spirit of flexibility


On Tue, Oct 8, 2019, 2:46 PM Tommy <notifications@github.com> wrote:

> This is very interesting. I'm not against using a different syntax for
> this.
> I wonder if in the future, we should focus on a single syntax instead of
> supporting more than one.
>
> Any reason why you prefer the bracket syntax in comparison to <>?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tleunen/pixi-multistyle-text/pull/78?email_source=notifications&email_token=ABRRWVJAFZVDSC4O72VFYHLQNSFLBA5CNFSM4HX4XJ4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEAUHCSY#issuecomment-539521355>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABRRWVPGNSQRSZ622MSLO7DQNSFLBANCNFSM4HX4XJ4A>
> .
>

blurymind commented:
Thank you for reviewing this. I will change it to enum and polish the or a bit more :)

tleunen commented:
gdevelop is a good argument ;) Please assign me as reviewer when it's ready and I'll make sure to spend some more time reviewing this.

blurymind commented:
@tleunen thank you :) 
I don't think I have the permission to assign reviewers to PRs on this tracker, but I can ping you with a comment when it's ready?

blurymind commented:
@tleunen I replaced the tag style declarations with enums :)
Can you review this? I think I am getting closer to getting my bbcode object ready for gdevelop. Its feature complete now

blurymind commented:
I don't mind :) I will add it at the end. Now when testing it extensively in gdevelop, I found a way to crash it - so I pushed a fix for that

Btw how did you make the html demo file? When I edit it in vscode,prettier moves the lines. I have to add it carefully with prettier turned off

tleunen commented:
It was probably before prettier 😅 You could use "Save without formatting" in vscode (or similar editors) to prevent prettier to run too.

blurymind commented:
updated demo now to show bbcode tags :) I added a couple of custom ones (spacing, outline and shadow).. I couldn't figure out a way to properly do the [u]underlined[/u] tag with pixi

blurymind commented:
btw once approved, you have to do the merge, since I have no permission to do merges on this tracker :)

tleunen commented on demo/index.html:
```diff
@@ -189,10 +199,11 @@ <h2>Have Fun</h2>
 renderer.backgroundColor = 0x333333;
 document.getElementById("pixi-container").appendChild(renderer.view);
 let stage = new PIXI.Container();
+const MultiStyleText = this.pixiMultistyleText;
```
Why is this needed? This is making the demo to fail on my machine. 

Not sure why the name would become `pixiMultistyleText`. The name used in microbundle is `MultiStyleText`.

blurymind replied:
ah sorry,I had to add it to get it to load on linux. I will remove it now

blurymind commented:
> Everything looks good, but before merging and releasing a new version. I'd like to better understand the reason why we use `this.pixiMultistyleText`.

For some reason the demo import wasnt working on linux for me. When doing console.log(this), I found the library available under that namespace for some reason. Anyways, I removed that line from the pr now, as it wasn't really a part  of the feature - just a temporary way for me to play with the demo

tleunen commented:
Thank you for this feature @blurymind!

blurymind commented:
Thank you for taking the time to review it and also for developing this @tleunen 
Gdevelop users will really benefit from this and we will hopefully see it in many more pixijs games out there

4ian commented:
Thanks @blurymind for working on this and @tleunen for reviewing it (and making pixi-multistyle-text in the first place :)). GDevelop users will be happy to be able to use this!
